api/alembic/versions/c04f22e31997_import_zones.py

```diff
@@ -0,0 +1,83 @@
+"""Import zones
+
+Revision ID: c04f22e31997
+Revises: 17b1c787f420
+Create Date: 2022-08-31 22:56:52.264112
+
+"""
+from typing import Final
+import tempfile
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.orm.session import Session
+import geoalchemy2
+from shapely.geometry import MultiPolygon, Polygon
+from shapely import wkb
+from app.utils import esri
+
+
+# revision identifiers, used by Alembic.
+revision = 'c04f22e31997'
+down_revision = '17b1c787f420'
+branch_labels = None
+depends_on = None
+
+
+shape_type_table = sa.Table('advisory_shape_types', sa.MetaData(),
+                            sa.Column('id', sa.Integer),
+                            sa.Column('name', sa.String))
+
+shape_table = sa.Table('advisory_shapes', sa.MetaData(),
+                       sa.Column('id', sa.Integer),
+                       sa.Column('external_identifier', sa.String),
+                       sa.Column('shape_type', sa.Integer),
+                       sa.Column('geom', geoalchemy2.Geometry))
+
+
+def upgrade():
+    session = Session(bind=op.get_bind())
+    statement = shape_type_table.insert().values(name='fire_zone').returning(shape_type_table.c.id)
+    result = session.execute(statement).fetchone()
+    shape_type_id = result.id
+
+    # We fetch a list of object id's, fetching the entire layer in one go, will most likely crash
+    # the server we're talking to.
+    zone_url: Final = "https://maps.gov.bc.ca/arcserver/rest/services/whse/bcgw_pub_whse_legal_admin_boundaries/MapServer/8"
+    zone_ids = esri.fetch_object_list(zone_url)
+    for object_id in zone_ids:
+        # Fetch each object in turn.
+        obj = esri.fetch_object(object_id, zone_url)
+        for feature in obj.get('features', []):
+            attributes = feature.get('attributes', {})
+            # Each zone is uniquely identified by a fire zone id.
+            mof_fire_zone_id = attributes.get('MOF_FIRE_ZONE_ID')
+            fire_zone_id = str(int(mof_fire_zone_id))
+            geometry = feature.get('geometry', {})
+            rings = geometry.get('rings', [[]])
```

Comment:
I think I missed something - what are these "rings"?

Reply:
For some reason that's the convention on the ARC format. No idea why they're using that convention. It's all over their documentation: https://totalapis.github.io/api-reference/esri-geometry-Polygon.html

Reply:
You want a comment in the code? It doesn't make it an less confusing....

Reply:
Their definition for "rings" is "an array of rings", which is not overly helpful, but regardless, yes I'd like that link commented in the code so at least it's documented that the terminology used there is a convention and not something you just pulled out of the dark recesses of your brain.

Reply:
one ring to rule them all, and in the darkness bind them